lib/ansible/modules/crypto/openssl_certificate.py

```diff
@@ -537,34 +537,35 @@ def _validate_keyUsage():
             if self.keyUsage:
                 for extension_idx in range(0, self.cert.get_extension_count()):
                     extension = self.cert.get_extension(extension_idx)
-                    if extension.get_short_name() == 'keyUsage':
+                    if extension.get_short_name() == b'keyUsage':
                         keyUsage = [OpenSSL._util.lib.OBJ_txt2nid(keyUsage) for keyUsage in self.keyUsage]
-                        current_ku = [OpenSSL._util.lib.OBJ_txt2nid(usage.strip()) for usage in str(extension).split(',')]
+                        current_ku = [OpenSSL._util.lib.OBJ_txt2nid(to_bytes(usage.strip())) for usage in str(extension).split(',')]
```

Comment:
Taking another look at this (and the similar code on line 555) there's a few things to change:
``` python
 current_ku = [OpenSSL._util.lib.OBJ_txt2nid(usage.strip()) for usage in to_bytes(extension, errors='surrogate_or_strict').split(b',')]
```

the changes I noticed:
* No need to use both str and to_bytes.  Use to_bytes once to save one encoding step.
* Use surrogate_or_strict so that we raise an exception if we can't encode something in utf-8.  The default error handler for to_bytes is surrogate_then_replace which could end up with replacement characters in the output.

Reply:
Line 568 is also similar

Reply:
`extension` comes from `extension = self.cert.get_extension(extension_idx)` which is an object, not a string or bytestring (https://pyopenssl.org/en/stable/api/crypto.html#OpenSSL.crypto.X509Extension). Pyopenssl itself usually just returns bytes, which is the reason `extension.get_short_name() == 'keyUsage' never will be True on python 3 (https://pyopenssl.org/en/stable/api/crypto.html#OpenSSL.crypto.X509Extension.get_short_name).

There are helper functions, like `__str__()` (https://pyopenssl.org/en/stable/api/crypto.html#OpenSSL.crypto.X509Extension.__str__) that serialize the contents of this extension which unfortunately can't really be nicely reached any other way unless (like on line 310 in openssl_csr.py in https://github.com/ansible/ansible/pull/30338) you manually work around bugs or not exposed functionality in pyopenssl directly. To me the implementation seems a bit arcane (https://github.com/pyca/pyopenssl/blob/master/src/OpenSSL/crypto.py#L761), so I'm happy to let the library handle this for me, even if it means that I would need to create a new bytestring from the result.

This means the `str(extension)` call is not optional and `to_bytes(extension, ...)` will probably not work, as this is not a string initially (maybe the __str__() method is called internally though when calling `to_bytes()`?).

Lastly there still could be some cleanup done in this module to make sure that stuff that is a string is used as a string and bytes stay bytes. I'm not a huge fan of the `_sanitize_inputs()` function that changes members in the class, not when they are used - but then again that's mostly a style argument and sooner or later this module could be rewritten using the cryptography module anyways I hope.